Preprocessing & Annotation functionality

dhensle opened the pull request:
This pull request standardizes preprocessing and annotation capabilities across all activitysim abm models.

### Refactoring of Preprocessing and Table Annotation

* **Standardized Preprocessing**:
  - Replaced custom preprocessing logic using `expressions.assign_columns` with `expressions.annotate_preprocessors` in models such as `auto_ownership`, `free_parking`, and `joint_tour_composition`. This change ensures consistency across models and reduces duplication. [[1]](diffhunk://#diff-97193db9ed95c8b33446406a13789cbd5570fd49a7b3f043ca12111fdb7ed68eL72-R75) [[2]](diffhunk://#diff-957afdcc6ef9a84be227a4f6ae091f2a430cb48128f7b6ffaa1d1855199562a1R86-R94) [[3]](diffhunk://#diff-6085d69d311dfc23c8ab7484c9aa5598253a2b850e6aa8fc872179fbe2cdd417L81-L110)

* **Unified Table Annotation**:
  - Replaced individual table annotation logic with `expressions.annotate_tables` in models like `atwork_subtour_destination`, `cdap`, and `joint_tour_frequency`. This provides a consistent and reusable approach for post-model table updates. [[1]](diffhunk://#diff-8f603b842d482c002976b2c1fba2a46107667e5306d1022705a768125241e0a6R123-R130) [[2]](diffhunk://#diff-803273e130a15732ee40d4de40bb6de04cd44bc28e6c0f0a044d87965123d88dL244-R279) [[3]](diffhunk://#diff-7532555815a91259f787dbe4557f79c98869f67e818ea2146c51050379a5f904R208-R215)

### Simplification of Model Settings

* **Removal of Unused Fields**:
  - Removed unnecessary fields like `preprocessor` and `annotate_*` from model settings classes such as `AtworkSubtourFrequencySettings`, `AutoOwnershipSettings`, and `FreeParkingSettings`. This cleanup eliminates unused or redundant configuration options. [[1]](diffhunk://#diff-8cfcad8139f5db1403922da2b25bd654cd57876d1d9e19347fdc32c1bd72b06bL40-R41) [[2]](diffhunk://#diff-97193db9ed95c8b33446406a13789cbd5570fd49a7b3f043ca12111fdb7ed68eL31-R30) [[3]](diffhunk://#diff-957afdcc6ef9a84be227a4f6ae091f2a430cb48128f7b6ffaa1d1855199562a1L29-L31)

* **Reorganization in `cdap`**:
  - Consolidated preprocessor and annotation settings into a single unified structure for better clarity and alignment with other models.

### Summary table of model changes:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/DAVID~1.HEN/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/DAVID~1.HEN/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{font-weight:700;}
-->

</head>

<body link="#0563C1" vlink="#954F72">


Model   name | Choosers Preprocessor | Alts Preprocessor | Annotator | Notes
-- | -- | -- | -- | --
accessibility | existed | N/A | N/A | model is essentially a preprocessor
atwork_subtour_destination | added | added | added | added preprocessing in both sample and simulate step in   util.tour_destination.py
atwork_subtour_frequency | updated | N/A | added |  
atwork_subtour_mode_choice | updated | N/A | added |  
atwork_subtour_scheduling | added | existed | added | ALTS_PREPROCESSOR segmented by tour purpose, added in   util.vectorize_tour_scheduling.py
auto_ownership | updated | N/A | added |  
cdap | added | N/A | updated |  
disaggregate_accessibility | existed | N/A | existed | now can use alts preprocessors in logsum calculations through tour and   location choice models
free_parking | updated | N/A | updated |  
initialize | N/A | N/A | updated | Model is annotator
initialize_los | N/A | N/A | N/A |  
initialize_tours | N/A | N/A | updated | Model is annotator
input_checker | N/A | N/A | N/A |  
joint_tour_composition | updated | N/A | updated |  
joint_tour_destination | added | added | added | added preprocessing in both sample and simulate step in   util.tour_destination.py
joint_tour_frequency | updated | N/A | added |  
joint_tour_frequency_composition | updated | updated | updated |  
joint_tour_participation | updated | N/A | added |  
joint_tour_scheduling | updated | added | updated | ALTS_PREPROCESSOR segmented by tour purpose, added in   util.vectorize_tour_scheduling.py
location_choice | added | added | updated | added preprocessing in both sample and simulate step
mandatory_scheduling | added | existed | added | ALTS_PREPROCESSOR segmented by tour purpose, added in   util.vectorize_tour_scheduling.py
mandatory_tour_frequency | updated | N/A | added |  
non_mandatory_destination | added | added | added | added preprocessing in both sample and simulate step in   util.tour_destination.py
non_mandatory_scheduling | added | added | added | ALTS_PREPROCESSOR segmented by tour purpose, added in   util.vectorize_tour_scheduling.py
non_mandatory_tour_frequency | updated | added | added |  
parking_location_choice | updated | added | added |  
school_escorting | existed | added | added | uniquely has 3 preprocessors
stop_frequency | updated | N/A | added |  
summarize | N/A | N/A | existed | Model is annotator
telecommute_frequency | updated | N/A | added |  
tour_mode_choice | updated | N/A | added |  
tour_od_choice | updated | added | added | added preprocessing in both sample and simulate step
tour_scheduling_probabilistic | N/A | N/A | added |  
transit_pass_ownership | updated | N/A | added |  
transit_pass_subsidy | updated | N/A | added |  
trip_departure_choice | existed | added | added |  
trip_destination | updated | updated | added | added preprocessing in both sample and simulate step
trip_matrices | existed | N/A | N/A |  
trip_mode_choice | existed | N/A | updated |  
trip_purpose | updated | N/A | added |  
trip_purpose_and_destination | N/A | N/A | added |  
trip_scheduling | updated | N/A | added | is probabilistic model
trip_scheduling_choice | updated | added | added |  
vehicle_allocation | updated | N/A | added |  
vehicle_type_choice | updated | updated | updated |  
work_from_home | updated | N/A | updated |  

aletzdy committed:
initial batch of models with pre and post processing added

aletzdy committed:
nonmand sched and transit pass models

dhensle committed:
first pass at preprocessing and annotate functionality in all models

dhensle committed:
fixing bugs in jtf and trip purpose

dhensle committed:
adding persons back in to locals_d in jtc

dhensle committed:
model name missing in tour scheduling

dhensle committed:
missing expressions import in tour sched prob

jpn-- commented:
A few thoughts before I dig into this more:

1. Estimation mode tests are failing.
2. Other than estimation mode, existing tests do show we have not broken anything that was tested. But this PR appears to add significant functionality, and lacks any testing of that new functionality.

dhensle commented:
Hey @jpn--, thanks for your first look. As mentioned during the call today, I have a few more things to clean up on the CI tests (which is why the PR is listed as a "draft").  I will ping you in the next few days when my commits are complete and this PR is ready for review.

dhensle committed:
ci unit tests & fixing estimation test error

dhensle commented on activitysim/abm/models/non_mandatory_tour_frequency.py:
```diff
@@ -233,27 +226,36 @@ def non_mandatory_tour_frequency(
-
+    # preprocessing choosers
     constants = config.get_model_constants(model_settings)
+    locals_dict = {
+        "person_max_window": lambda x: person_max_window(state, x),
+        "person_available_periods": lambda persons, start_bin, end_bin, continuous: person_available_periods(
+            state, persons, start_bin, end_bin, continuous
+        ),
+    }
+    locals_dict.update(constants)
```
rejecting cause its not a good suggestion.

dhensle commented:
@jpn-- Ok, this is ready for your review.  I fixed the estimation mode issue and added unit tests for both the preprocessor and annotation functions which include some common expressions including reindex, skims lookups (in multiple ways), and some groupby calculations across tables.  If you have other CI test ideas, I am happy to hear them.

jpn-- commented on activitysim/abm/models/vehicle_allocation.py:
```diff
@@ -90,20 +90,15 @@ class VehicleAllocationSettings(LogitComponentSettings, extra="forbid"):
-    """Setting for the preprocessor."""
-
-    OCCUPANCY_LEVELS: list = [1]  # TODO Check this
+    OCCUPANCY_LEVELS: list = [1, 2, 3.5]
```
Clearly, the previous default value here of [1] was not a great solution.  But it seems the occupancy levels set here needs to jive with calls to the generated variable names set in spec files for other components.  "df.vehicle_occup_1" is going to show up as a variable name elsewhere in other components.  I think it would be better to make that clear in the docstring, and have no default at all (i.e. failure to define this in the config for this model is an error).

jpn-- replied:
Also, this appears to be basically unrelated to the fundamental nature of this PR, maybe just leave it out here, and we can address it in a separate specific PR for this.

dhensle replied:
Sure, we can address this separately.  Reverted.

jpn-- commented on activitysim/abm/models/mandatory_tour_frequency.py:
`MandatoryTourFrequencySettings` still has a redundant `annotate_persons` defined, which it also inherits from `LogitComponentSettings`

dhensle replied:
Good catch, removed.

jpn-- commented on activitysim/abm/models/cdap.py:
```diff
@@ -36,12 +36,16 @@ class CdapSettings(PydanticReadable, extra="forbid"):
 
+    preprocessor: PreprocessorSettings | None = None
+    """Preprocess choosers tables before running the model."""
+    annotate_persons: PreprocessorSettings | None = None
```
Automatic documentation tools will work better if `annotate_persons` has its own docstring, even though it's clear to a human reader the one docstring below applies to both.

dhensle replied:
done

jpn-- commented on activitysim/core/configuration/logit.py:
```diff
@@ -77,6 +77,27 @@ class BaseLogitComponentSettings(PydanticCompute):
+    utility expressions.
+    """
+
+    annotate_households: PreprocessorSettings | None = None
```
For autodoc, have each annotate_* field have its own docstring.

dhensle replied:
done

jpn-- commented on activitysim/core/util.py:
```diff
@@ -704,9 +704,9 @@ def drop_unused_columns(
         unique_variables_in_spec.update(re.findall(pattern, custom_chooser_lines))
 
-    logger.info("Dropping unused variables in chooser table")
+    logger.debug("Dropping unused variables in chooser table")
```
This is clearly unrelated to this PR.

I do think we should address our logging, by default we are logging far too much

dhensle replied:
yeah, I put this in my code cause it was annoying to look at while debugging. Happy to address this in a separate logging task.  reverted.

jpn-- reviewed (changes requested):
This is great!  I found very little substantive to address, just a few things to clean up.  Thanks for all the hard work on this.

dhensle committed:
addressing review comments

dhensle commented:
@jpn-- thanks for your comments.  I have addressed them, but please let me know if there's anything else you find!
